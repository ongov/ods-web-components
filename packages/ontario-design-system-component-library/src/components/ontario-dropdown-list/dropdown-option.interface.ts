export interface DropdownOption {
  /**
   * The text to display as label.
   *
   * @example
   * <ontario-dropdown-list label="Do you like cats?" name="cat-dropdown" required
   * is-empty-start-option="Please select" options='[{
   *     "value": "dropdown-list-1",
   *     "label": "Option 1"
   *   },
   *   {
   *     "value": "dropdown-list-2",
   *     "label": "Option 2"
   *   },
   *   {
   *      "value": "dropdown-list-3",
   *      "label": "Option 3"
   *   }]'>
   * </ontario-dropdown-list>
   *
   * The resulting text for the first dropdown option will show `"Option 1"`.
   * The resulting text for the first dropdown option will show `"Option 2"`.
   * The resulting text for the first dropdown option will show `"Option 3"`.
   */
  label: string;

  /**
   * The dropdown option content value.
   * Each value must be unique to the option.
   */
  value: string;
}
