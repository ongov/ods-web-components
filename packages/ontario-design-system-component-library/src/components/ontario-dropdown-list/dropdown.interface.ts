import { Base } from '../../utils/common.interface';
import { DropdownOption } from './dropdown-option.interface';

export interface Dropdown extends Base {
  /**
   * The legend for the Dropdown List.
   */
  legend: string;

  /**
   * The name for the dropdown list.
   */
  name: string;

  /**
   * The ID for the dropdown list.
   */
  elementId?: string;

  /**
   * Each property will be passed in through an object in the options array.
   * This can either be passed in as an object directly (if using react), or as a string in HTML.
   * In the example below, the options are being passed in as a string and
   * there are three dropdown options to be displayed in the fieldset.
   *
   * @example
   * <ontario-dropdown-list legend="Do you like cats?" name="cat-dropdown" is-required
   * is-empty-start-option="Please select" options='[{
   *     "value": "dropdown-list-1",
   *     "label": "Option 1"
   *   },
   *   {
   *     "value": "dropdown-list-2",
   *     "label": "Option 2"
   *   },
   *   {
   *      "value": "dropdown-list-3",
   *      "label": "Option 3"
   *   }]'>
   * </ontario-dropdown-list>
   */
  options: string | DropdownOption[];

  /**
   * Determine whether the dropdown list is required.
   * If required, add `is-required` attribute.
   * Otherwise, the `optional` flag will appear.
   *
   * @example
   * <ontario-dropdown-list ... is-required></ontario-dropdown-list>
   */
  isRequired: boolean;

  /**
   * Whether or not the initial option displayed is empty.
   * If set to true, it will render the default “select” text.
   * If set to a string, it will render the string value.
   *
   * @example
   * <ontario-dropdown-list is-empty-start-option></ontario-dropdown-list>
   *
   * or
   *
   * <ontario-dropdown-list is-empty-start-option="Please select"></ontario-dropdown-list>
   */
  isEmptyStartOption?: boolean | string;
}
