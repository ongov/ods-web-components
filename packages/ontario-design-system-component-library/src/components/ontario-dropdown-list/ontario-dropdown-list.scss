@import '../../global.scss';

.ontario-dropdown {
  background-color: $ontario-colour-white;
  background-position: right $spacing-2 center;
  background-repeat: no-repeat;
  background-size: $spacing-6;
  padding-right: calc(#{$spacing-4}* 2.25);
  cursor: pointer;
}

/* These styles are to fix differences in default styling for dropdowns in IE */
.ontario-dropdown::-ms-expand {
  display: none;
}

.ontario-dropdown option {
  font-weight: normal;
}
