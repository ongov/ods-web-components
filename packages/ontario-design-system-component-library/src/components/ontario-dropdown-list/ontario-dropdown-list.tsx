import { Component, State, Element, h, Prop, Listen, Watch, getAssetPath } from '@stencil/core';
import { InputCaption } from '../../utils/input-caption/input-caption';
import { Caption } from '../../utils/input-caption/caption.interface';
import { DropdownOption } from './dropdown-option.interface';
import { Dropdown } from './dropdown.interface';
import { v4 as uuid } from 'uuid';
import { validateObjectExists, validatePropExists } from '../../utils/validation/validation-functions';
import { ConsoleMessageClass } from '../../utils/console-message/console-message';
import { default as translations } from '../../translations/global.i18n.json';

@Component({
	tag: 'ontario-dropdown-list',
	styleUrl: 'ontario-dropdown-list.scss',
	shadow: true,
	assetsDirs: ['./assets'],
})
export class OntarioDropdownList implements Dropdown {
	/**
	 * Grant access to the host element and related DOM methods/events within the class instance.
	 */
	@Element() element: HTMLElement;

	/**
	 * The text to display as the label
	 *
	 * @example
	 * <ontario-dropdown-list
	 *   caption='{
	 *     "captionText": "Address",
	 *     "captionType": "heading",
	 *   }'
	 *   required="true"
	 *   ...>
	 * </ontario-dropdown-list>
	 */
	@Prop() caption: Caption | string;

	/**
	 * The language of the component.
	 * This is used for translations, and is by default set through event listeners checking for a language property from the header. If none is passed, it will default to English.
	 */
	@Prop({ mutable: true }) language?: string = 'en';

	/**
	 * The name for the dropdown list.
	 */
	@Prop() name: string;

	/**
	 * The ID for the dropdown list.
	 * If no ID is provided, one will be generated.
	 */
	@Prop({ mutable: true }) elementId?: string;

	/**
	 * Each property will be passed in through an object in the options array.
	 * This can either be passed in as an object directly (if using react), or as a string in HTML.
	 * In the example below, the options are being passed in as a string and
	 * there are three dropdown options to be displayed in the fieldset.
	 *
	 * @example
	 * <ontario-dropdown-list
	 *   caption='{
	 *     "captionText": "Do you like cats?",
	 *     "captionType": "heading",
	 *   }'
	 *   name="cat-dropdown"
	 *   is-empty-start-option="Please select"
	 *   options='[{
	 *     "value": "dropdown-list-1",
	 *     "label": "Option 1"
	 *   },
	 *   {
	 *     "value": "dropdown-list-2",
	 *     "label": "Option 2"
	 *   },
	 *   {
	 *      "value": "dropdown-list-3",
	 *      "label": "Option 3"
	 *   }]'
	 *   required="true"
	 * >
	 * </ontario-dropdown-list>
	 */
	@Prop() options: string | DropdownOption[];

	/**
	 * This is used to determine whether the dropdown list is required or not.
	 * This prop also gets passed to the InputCaption utility to display either an optional or required flag in the label.
	 * If no prop is set, it will default to false (optional).
	 */
	@Prop() required?: boolean = false;

	/**
	 * Whether or not the initial option displayed is empty.
	 * If set to true, it will render the default “select” text.
	 * If set to a string, it will render the string value.
	 *
	 * @example
	 * <ontario-dropdown-list is-empty-start-option></ontario-dropdown-list>
	 *
	 * or
	 *
	 * <ontario-dropdown-list is-empty-start-option="Please select"></ontario-dropdown-list>
	 */
	@Prop() isEmptyStartOption?: boolean | string = false;

	/**
	 * Instantiate an InputCaption object for internal logic use
	 */
	@State() private captionState: InputCaption;

	/**
	 * The options are re-assigned to the internalOptions array.
	 */
	@State() private internalOptions: DropdownOption[];

	/**
	 * This listens for the `setAppLanguage` event sent from the test language toggler when it is is connected to the DOM. It is used for the initial language when the input component loads.
	 */
	@Listen('setAppLanguage', { target: 'window' })
	handleSetAppLanguage(event: CustomEvent<string>) {
		this.language = event.detail;
	}

	@Listen('headerLanguageToggled', { target: 'window' })
	handleHeaderLanguageToggled(event: CustomEvent<string>) {
		const toggledLanguage = event.detail;
		this.language = toggledLanguage;
	}

	/*
	 * Watch for changes in the `name` prop for validation purpose
	 * Validate the name and make sure the name has a value.
	 * Log warning if user doesn't input a value for the name.
	 */
	@Watch('name')
	validateName(newValue: string) {
		if (validatePropExists(newValue)) {
			const message = new ConsoleMessageClass();
			message.addDesignSystemTag().addMonospaceText(' name ').addRegularText('for').addMonospaceText(' <ontario-dropdown-list> ').addRegularText('was not provided').printMessage();
		}
	}

	/*
	 * Watch for changes in the `options` prop for validation purpose
	 * Validate the options and make sure the options has a value.
	 * Log warning if user doesn't input a value for the options.
	 */
	@Watch('options')
	validateOptions(newValue: object) {
		if (validateObjectExists(newValue)) {
			const message = new ConsoleMessageClass();
			message
				.addDesignSystemTag()
				.addMonospaceText(' options ')
				.addRegularText('for')
				.addMonospaceText(' <ontario-dropdown-list> ')
				.addRegularText('was not provided')
				.printMessage();
		}
	}

	@Watch('options')
	parseOptions() {
		if (typeof this.options !== 'undefined') {
			if (!Array.isArray(this.options)) {
				this.internalOptions = JSON.parse(this.options);
			} else {
				this.internalOptions = this.options;
			}
		}
	}

	@Watch('caption')
	private updateCaptionState(newValue: Caption | string) {
		this.captionState = new InputCaption(this.element.tagName, newValue, translations, this.language, false, this.required);
	}

	/**
	 * Watch for changes in the `language` to render either the English or French translations
	 */
	@Watch('language')
	updateLanguage() {
		this.updateCaptionState(this.caption);
	}

	public getId(): string {
		return this.elementId ?? '';
	}

	private getDropdownArrow() {
		return {
			backgroundImage: `url(${getAssetPath('./assets/ontario-material-dropdown-arrow-48px.svg')})`,
		};
	}

	componentWillLoad() {
		this.updateCaptionState(this.caption);
		this.parseOptions();
		this.validateName(this.name);
		this.validateOptions(this.internalOptions);
		this.elementId = this.elementId ?? uuid();
	}

	render() {
		return (
			<div class="ontario-form-group">
				{this.captionState.getCaption(this.getId())}
				<select class="ontario-input ontario-dropdown" id={this.getId()} name={this.name} style={this.getDropdownArrow()} required={!!this.required}>
					{this.isEmptyStartOption && (this.isEmptyStartOption === true ? <option>Select</option> : <option>{this.isEmptyStartOption}</option>)}

					{this.internalOptions?.map(dropdown => <option value={dropdown.value}>{dropdown.label}</option>) ?? ''}
				</select>
			</div>
		);
	}
}
