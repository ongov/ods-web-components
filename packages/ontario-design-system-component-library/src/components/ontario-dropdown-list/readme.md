# ontario-dropdown-list

Only use a dropdown (select) list if you cannot use other form components to capture the user’s information.

## Usage Guidance

Please refer to the [Ontario Design System](https://designsystem.ontario.ca/components/detail/dropdown-lists.html) for current documentation guidance.

## Configuration

Once the component package has been installed (see Ontario Design System Component Library for installation instructions), the dropdown-list component can be added directly into the project's code, and can be customized by updating the properties outlined [here](#properties). Additional information on custom types for dropdown list properties are outlined [here](#custom-property-types). Please see the [examples](#examples) below for how to configure the component.

## Examples

Example of a dropdown list component with no `elementId` or `isEmptyStartOption` props passed. Note that by default, an ID for the `elementId` will be generated if none is provided. If no `isEmptyStartOption` prop is provided, it will default to `false` and display the first option in the `options` list.

```html
<ontario-dropdown-list
	name="streaming-service"
	caption='{
			"captionText": "Select a streaming service",
			"captionType": "default"
		}'
	required="true"
	options='[{
			"value": "netflix",
			"label": "Netflix"
		},
		{
			"value": "disney-plus",
			"label": "Disney Plus"
		},
		{
			"value": "crave",
			"label": "Crave"
		},
		{
			"value": "prime",
			"label": "Prime Video"
		}]'
>
</ontario-dropdown-list>
```

In the following example, all available props are passed through.

```html
<ontario-dropdown-list
	name="provinces-territories"
	is-empty-start-option="Select"
	element-id="provinces-territories"
	language="en"
	caption='{
			"captionText": "Province/territory",
			"captionType": "heading"
		}'
	required="true"
	options='[{
			"value": "alberta",
			"label": "Alberta"
		},
		{
			"value": "british-columbia",
			"label": "British Columbia"
		},
		{
			"value": "manitoba",
			"label": "Manitoba"
		},
		{
			"value": "new-brunswick",
			"label": "New Brunswick"
		},
		{
			"value": "newfoundland-and-labrador",
			"label": "Newfoundland and Labrador"
		},
		{
			"value": "nova-scotia",
			"label": "Nova Scotia"
		},
		{
			"value": "ontario",
			"label": "Ontario"
		},
		{
			"value": "prince-edward-island",
			"label": "Prince Edward Island"
		},
		{
			"value": "quebec",
			"label": "Quebec"
		},
		{
			"value": "saskatchewan",
			"label": "Saskatchewan"
		},
		{
			"value": "northwest-territories",
			"label": "Northwest Territories"
		},
		{
			"value": "nunavut",
			"label": "Nunavut"
		},
		{
			"value": "yukon",
			"label": "Yukon"
		}]'
>
</ontario-dropdown-list>
```

## Custom property types

### Caption

```html
caption='{ "captionText": "Province/territory", "captionType": "heading" }'
```

| **Property name** | **Type**                           | **Description**                                                                                         |
| ----------------- | ---------------------------------- | ------------------------------------------------------------------------------------------------------- |
| `captionText`     | `string`                           | The text to display as the caption (label) for the dropdown-list. This must be implemented.             |
| `captionType`     | `"default" \| "large"\| "heading"` | The type of caption to render. This is optional - if no type is passed, the "default" type will render. |

### Options

```html
options='[ { "value": "netflix", "label": "Netflix" }, { "value": "disney-plus", "label": "Disney Plus" }, { "value":
"crave", "label": "Crave" }, { "value": "prime", "label": "Prime Video" } ]'
```

| **Property name** | **Type** | **Description**                                                             |
| ----------------- | -------- | --------------------------------------------------------------------------- |
| `value`           | `string` | The dropdown option content value. Each value must be unique to the option. |
| `label`           | `string` | The text to display as the dropdown option label.                           |

## Accessibility

- An `id` attribute is necessary to allow the dropdown list to be associated with a label element. If none is provided through the `elementId` prop, one will be generated.
- A `name` attribute needs to be set to be submitted to the server when the form is submitted.
- Do not add any customized styles to dropdown lists - the browser's default is the most accessible.

<!-- Auto Generated Below -->

## Properties

| Property             | Attribute               | Description                                                                                                                                                                                                                                                                                                  | Type                                  | Default     |
| -------------------- | ----------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | ------------------------------------- | ----------- |
| `caption`            | `caption`               | The text to display as the label                                                                                                                                                                                                                                                                             | `Caption \| string`                   | `undefined` |
| `elementId`          | `element-id`            | The ID for the dropdown list. If no ID is provided, one will be generated.                                                                                                                                                                                                                                   | `string \| undefined`                 | `undefined` |
| `hintExpander`       | `hint-expander`         | Used to include the Hint Expander component underneath the dropdown list box. This is passed in as an object with key-value pairs. This is optional.                                                                                                                                                         | `HintExpander \| string \| undefined` | `undefined` |
| `hintText`           | `hint-text`             | Hint text for Ontario Dropdown. This is optional.                                                                                                                                                                                                                                                            | `string \| undefined`                 | `undefined` |
| `isEmptyStartOption` | `is-empty-start-option` | Whether or not the initial option displayed is empty. If set to true, it will render the default “select” text. If set to a string, it will render the string value.                                                                                                                                         | `boolean \| string \| undefined`      | `false`     |
| `language`           | `language`              | The language of the component. This is used for translations, and is by default set through event listeners checking for a language property from the header. If none is passed, it will default to English.                                                                                                 | `"en" \| "fr" \| undefined`           | `'en'`      |
| `name`               | `name`                  | The name for the dropdown list.                                                                                                                                                                                                                                                                              | `string`                              | `undefined` |
| `options`            | `options`               | Each property will be passed in through an object in the options array. This can either be passed in as an object directly (if using react), or as a string in HTML. In the example below, the options are being passed in as a string and there are three dropdown options to be displayed in the fieldset. | `DropdownOption[] \| string`          | `undefined` |
| `required`           | `required`              | This is used to determine whether the dropdown list is required or not. This prop also gets passed to the InputCaption utility to display either an optional or required flag in the label. If no prop is set, it will default to false (optional).                                                          | `boolean \| undefined`                | `false`     |

## Dependencies

### Depends on

- [ontario-hint-text](../ontario-hint-text)
- [ontario-hint-expander](../ontario-hint-expander)

### Graph

```mermaid
graph TD;
  ontario-dropdown-list --> ontario-hint-text
  ontario-dropdown-list --> ontario-hint-expander
  ontario-hint-expander --> ontario-icon-chevron-up
  ontario-hint-expander --> ontario-icon-chevron-down
  style ontario-dropdown-list fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
