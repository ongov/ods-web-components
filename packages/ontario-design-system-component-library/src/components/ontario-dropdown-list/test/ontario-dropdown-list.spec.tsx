import { newSpecPage } from '@stencil/core/testing';
import { OntarioDropdownList } from '../ontario-dropdown-list';

describe('ontario-dropdown-list', () => {
  describe('render', () => {
    it('should render a dropdown list element', async () => {
      const page = await newSpecPage({
        components: [OntarioDropdownList],
        html: `<ontario-dropdown-list element-id="dropdown-list-1"></ontario-dropdown-list>`,
      });
      expect(page.root).toEqualHtml(`
        <ontario-dropdown-list element-id="dropdown-list-1">
          <mock:shadow-root>
            <div class="ontario-form-group">
              <label class="ontario-label" htmlFor="ontario-dropdown-list">
                <span class="ontario-label__flag">
                  (optional)
                </span>
              </label>
              <select class="ontario-dropdown ontario-input" id="dropdown-list-1" style="background-image: url(/assets/ontario-material-dropdown-arrow-48px.svg);"></select>
            </div>
          </mock:shadow-root>
        </ontario-dropdown-list>
      `);
    });


    it('should reflect attributes/props being set', async () => {
      const page = await newSpecPage({
        components: [OntarioDropdownList],
        html: `<ontario-dropdown-list
                  label="ontario dropdown label"
                  name="dropdown-options"
                  element-id="dropdown-list-1",
                  is-empty-start-option="Please select"
								></ontario-dropdown-list>`,
      });

      expect(page.rootInstance.label).toBe('ontario dropdown label');
      expect(page.rootInstance.name).toBe('dropdown-options');
      expect(page.rootInstance.elementId).toBe('dropdown-list-1');
      expect(page.rootInstance.isEmptyStartOption).toBe("Please select");
    });
  });
});
