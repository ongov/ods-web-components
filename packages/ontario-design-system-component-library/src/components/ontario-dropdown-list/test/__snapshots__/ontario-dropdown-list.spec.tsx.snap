// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ontario-dropdown-list snapshot should render the expected html 1`] = `
<ontario-dropdown-list is-empty-start-option="Please select" is-required="" label="Do you like cats?" name="cat-dropdown" options="[{
          &quot; value&quot;: &quot;dropdown-list-1&quot;,
          &quot;label&quot;: &quot;Option 1&quot;
        },
        {
          &quot;value&quot;: &quot;dropdown-list-2&quot;,
          &quot;label&quot;: &quot;Option 2&quot;
        },
        {
          &quot;value&quot;: &quot;dropdown-list-3&quot;,
          &quot;label&quot;: &quot;Option 3&quot;
        }]">
  <mock:shadow-root>
    <div class="ontario-form-group">
      <label class="ontario-label" htmlfor="ontario-dropdown-list">
        Do you like cats?
        <span class="ontario-label__flag">
          (required)
        </span>
      </label>
      <select class="ontario-dropdown ontario-input" name="cat-dropdown" style="background-image: url(/assets/ontario-material-dropdown-arrow-48px.svg);">
        <option>
          Please select
        </option>
        <option>
          Option 1
        </option>
        <option value="dropdown-list-2">
          Option 2
        </option>
        <option value="dropdown-list-3">
          Option 3
        </option>
      </select>
    </div>
  </mock:shadow-root>
</ontario-dropdown-list>
`;
